lib/srv/desktop/deskproto/proto.go

```diff
@@ -0,0 +1,254 @@
+// Package deskproto implements the desktop protocol encoder/decoder.
+// See https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md
+//
+// TODO(awly): complete the implementation of all messages, even if we don't
+// use them yet.
+package deskproto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"image"
+	"image/png"
+	"io"
+
+	"github.com/gravitational/trace"
+)
+
+// MessageType identifies the type of the message.
+type MessageType byte
+
+const (
+	TypeClientScreenSpec = MessageType(1)
+	TypePNGFrame         = MessageType(2)
+	TypeMouseMove        = MessageType(3)
+	TypeMouseButton      = MessageType(4)
+	TypeKeyboardButton   = MessageType(5)
+	TypeClientUsername   = MessageType(7)
+)
+
+// Message is a Go representation of a desktop protocol message.
+type Message interface {
+	Encode() ([]byte, error)
+}
+
+// Decode decodes the wire representation of a message.
+func Decode(buf []byte) (Message, error) {
+	if len(buf) == 0 {
+		return nil, trace.BadParameter("input desktop protocol message is empty")
+	}
+	switch buf[0] {
+	case byte(TypeClientScreenSpec):
+		return decodeClientScreenSpec(buf)
+	case byte(TypePNGFrame):
+		return decodePNGFrame(buf)
+	case byte(TypeMouseMove):
+		return decodeMouseMove(buf)
+	case byte(TypeMouseButton):
+		return decodeMouseButton(buf)
+	case byte(TypeKeyboardButton):
+		return decodeKeyboardButton(buf)
+	case byte(TypeClientUsername):
+		return decodeClientUsername(buf)
+	default:
+		return nil, trace.BadParameter("unsupported desktop protocol message type %d", buf[0])
+	}
+}
+
+// PNGFrame is the PNG frame message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#2---png-frame
+type PNGFrame struct {
+	Img image.Image
+}
+
+func (f PNGFrame) Encode() ([]byte, error) {
+	type header struct {
+		Type          byte
+		Left, Top     uint32
+		Right, Bottom uint32
+	}
+
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.BigEndian, header{
+		Type:   byte(TypePNGFrame),
+		Left:   uint32(f.Img.Bounds().Min.X),
+		Top:    uint32(f.Img.Bounds().Min.Y),
```

Comment:
Shouldn't "top" be Max.Y and "bottom" - Min.Y?

Reply:
Desktop coordinate systems typically go from top-left (min) to bottom-right (max).
I think that mimics the order of how screens light up the pixels.